content/blog/2019-11-13-llvm-profiling/index.md

```diff
@@ -0,0 +1,336 @@
+    createInstr(bb, bbCounter, basic_block);
+    createInstr(bb, instrCounter,instr);
+    
+    return true;
+}
+
+```
+
+Finally, we print the profiling results by first create a function `printf` by specifying input argument and output argument types. Then transverse the `main` function. Right before `return` of the last block in `main`, create string variable corresponding to variables we are tracing and get the pointer to that string variable. Also, get the number variable by first get the pointer pointing to the global variable and then create a load instruction to get the actual pointer to the value. Finally, we can call printf function with input arguments.
+
+```c++
+    RBuilder<> Builder(M.getContext());
+    Function *mainFunc = M.getFunction("main")
+    // not the main module
+    if (!mainFunc)
+        return false;
+
+    // build printf function handle
+    std::vector<Type *> FTyArgs;
+    FTyArgs.push_back(Type::getInt8PtrTy(M.getContext())); // specify the first argument, i8* is the return type of CreateGlobalStringPtr
+    FunctionType *FTy = FunctionType::get(Type::getInt32Ty(M.getContext()), FTyArgs, true); // create function type with return type, argument types and if const argument 
+    FunctionCallee printF = M.getOrInsertFunction("printf", FTy); // create function if not extern or defined
+    
+    for (auto bb = mainFunc->begin(); bb != mainFunc->end(); bb++) {
+        for(auto it = bb->begin(); it != bb->end(); it++) {
+            if ((std::string)it->getOpcodeName() == "ret") {
+                // insert printf at the end of main function, before return function
+                Builder.SetInsertPoint(&*bb, it);
+                for(int i = 0; i < atomicCounter.size(); i++){
+                    Value *format_long = Builder.CreateGlobalStringPtr(atomicCounter[i]+": %ld\n", "formatLong");// create global string variable formatLong, add suffix(.1/.2/...) if already exists
+                    std::vector<Value *> argVec;
+                    argVec.push_back(format_long); 
+                    Value *atomic_counter = M.getGlobalVariable(atomicCounter[i]);// get pointer pointing to the global variable name
+                    Value* finalVal = new LoadInst(atomic_counter, atomic_counter->getName()+".val", &*it);// atomic_counter only points to a string, but we want to print the number the string stores
+                    argVec.push_back(finalVal); 
+                    CallInst::Create(printF, argVec, "printf", &*it); //create printf function with the return value name called printf (with suffix if already exists)
+                }
+            }
+        }
+    }
+    return true;
+}
+```
+
+
+
+### Expensive Operations Information
+
+The whole part follows the same flow as basic execution information except we need to distinguish the instruction type and increment corresponding counter on block basis. Therefore in each `runOnBasicBlock` method, we need the following lines:
```

Comment:
- I'm not sure what "The whole part" refers to… maybe "this part"?
- on _a_ block basis